### ISSUE TITLE
Сборка из исходных кодов

### ISSUE THREAD

**Issue by DrNorfolk** [xrip/pico-gameboy#2]
Привет авторам)
Спасибо за отличный проект! 
Я скачал готовый релиз этого репозитория и pico-megadrive, и они отлично работают. Однако, когда я пытаюсь собрать это проекты из исходников, то получаю неработающие прошивки.
Пробовал собирать под Windows и Debian, получаю разные результаты, но оба не работают:
- под Windows использую Pico SDK v1.5.1, cmake version 3.31.2 и GCC 10.3.1 arm-none-eabi, получаю стартовый синий экран с выбором игр, но при выборе рома и нажатии на кнопки START, всё зависает.
- под Debian использую  Pico SDK v2.1.0 , cmake version 3.25.1 и arm-none-eabi-gcc (15:12.2.rel1-1) 12.2.1 20221205, получаю черный экран.
Cmake запускаю с параметром -DTFT=ON, а затем make -j4

ps. Так же в CMakeLists.txt в target_link_libraries нужно добавить hardware_flash, потому как без него компилятор выдает:
fatal error: hardware/flash.h: No such file or directory

Спасибо за помощь!

**Comment by xrip** [comment]
Речь идет об pico-megadrive или об pico-gameboy? 
Пика зеленая, оригинальная или черная китайская? Все прошивки достаточно сильно разогнаны, возможно не заводится из-за этого. Кроме того, надо собирать в MinSizeRel а не в Release, во всяком случае, я собираю так.

**Comment by DrNorfolk** [comment]
Большое спасибо за оперативность!

Проверил, действительно, в режиме **MinSizeRel** под **Windows** сборка проходит успешно и прошивка работает. Отлично! Проверил и **pico-megadrive**, и **pico-gameboy**.
Под Linux, с описанным выше набором инструментов, пробую делать
`# cmake -DCMAKE_BUILD_TYPE=MinSizeRel -DTFT=ON ..`
затем 
**# make -j4**,
**pico-gameboy** собирается, но выдает черный экран, а размер прошивки сильно меньше, чем в сборках под Windows.
**pico-megadrive** под Linux не собирается, компилятор выдает 
```
In file included from /home/user/pico_projects/pico-megadrive/src/gwenesis/sound/gwenesis_sn76489.c:37:
/home/user/pico-sdk/src/rp2040/pico_platform/include/pico/platform.h:20:2: error: #error pico/platform.h should not be included directly; include pico.h instead
   20 | #error pico/platform.h should not be included directly; include pico.h instead
```
Здесь возможно отличие версии компилятора и pico-sdk влияет, продолжаю исследование этого вопроса.

Плату Pico использую фиолетовую китайскую, есть еще зеленая китайская, но на ней пока не проверял.
В меню, по нажатию SELECT+START в строке **Overclocking** стоит 378 MHz, пробовал увеличивать, но это приводит к зависанию.

В качестве пожеланий:
- можно добавить в описание проекта, что сборку нужно делать в **MinSizeRel**, а для выбора варианта видео-выхода  в Windows+VSCode+Cmake Tools+pico-sdk можно создать в корне проекта 
`.vscode\settings.json` с содержимым, например для TFT-дисплея:
```
{
    "cmake.configureSettings": {
        "TFT":true
    }
}
```
- добавить описание для сборки под Линукс, если такое тестировалось. Я попробую разные версии cmake и arm-none-eabi-gcc, вдруг получится найти подходящие, напишу, если будут результаты.
- в `CMakeLists.txt` для **pico-gameboy** в последнем **target_link_libraries** добавить указание на _hardware_flash_, без него совсем не компилируется проект.

**Comment by xrip** [comment]
```
можно добавить в описание проекта, что сборку нужно делать в MinSizeRel, а для выбора варианта видео-выхода в Windows+VSCode+Cmake Tools+pico-sdk можно создать в корне проекта
.vscode\settings.json с содержимым, например для TFT-дисплея:
```
Для написания кода я использую Jetbrains Clion, так что тоже врядли помогу, но с удовольствием приму пул-реквест.
Список доступных опций сборки:
```CMAKE
-DHDMI
-DVGA
-DTFT # ST7789
-DTFT -DILI9341 # ILI9341
-DTV # RGBi Scart
-DSOFT-TV # Composite TV
-DI2S # I2S sound DAC
```

Сборки под Linux не тестировались и не поддерживаются, как минимум потому что если человек использует линукс, он вполне технически подкован, чтобы решить все проблемы :)

Pico-gameboy долгое время собирался Github actions и вполне нормально работал:
https://github.com/xrip/pico-gameboy/blob/main/.github/workflows/build-release-on-tag.yml

**Comment by DrNorfolk** [comment]
> если человек использует линукс, он вполне технически подкован, чтобы решить все проблемы :)

Сборкой под Линукс я занялся только потому, что под Windows не получилось собрать, подумал, возможно моя система и инструменты не подходят.

В любом случае, спасибо за проект и помощь)

И пользуясь случаем, хотел еще два момента узнать:
- есть ли в обозримых планах появление в проекте возможности подключения внешних кнопок, вроде того, как уже спрашивали в репозитории с [pico-megadrive](https://github.com/xrip/pico-megadrive/issues/3#issue-2614098407) или стоит по совету из [этого ](https://github.com/xrip/pico-megadrive/issues/3#issuecomment-2440190175) сообщения заняться самостоятельной разработкой такой возможности, а этот проект предназначен только под существующее железо Мурмулятора?
- можно ли изменить разрешение для TFT-дисплея, чтобы картинка была "квадратной" как у оригинального Gameboy. Сейчас при сборке получается 320x240  (и у дисплея такое же разрешение) и изображение растягивается, сплющивается, а по верхнему и нижнему краю изображение отрезается.

**Comment by xrip** [comment]
Внешние кнопки поддерживаться не будут, и самый простой способ - это купить денди джойстик и использовать его как плату управления внешними кнопками. Или реализовать свою схему на 4021. Быстро, дешево и сердито. Какая разница -- провода тянуть к пике или к этой плате?

Можно, только ТФТ экраны в которые можно вписать целочисленно умножив 160х144 -- либо не существуют, либо стоят как геймбой в отличном состоянии.
Самое близкое что можно использовать -- это экран 480x320 перевернув его в портретный режим и использовать чуть больше половины по высоте, использовав множитель 2, получив таким образом разрешение 320х288, так, к слову сказать делают в IPS мод для оригинального геймбоя.

**Comment by DrNorfolk** [comment]
> Или реализовать свою схему на 4021. Быстро, дешево и сердито

Да, действительно, для Геймбоя такая схема вполне себе подойдет. Тоже подумал, что можно будет реализовать такую плату и всё упаковать в корпус Геймбоя.
А вот подключение оригинального/клона-джойстика Сега для **pico-megadrive** - такое уже не пройдет, либо совсем такой джойстик переделывать, оставив от него лишь корпус.

Про TFT понял, спасибо за информацию. Буду пока пользоваться таким "растянутым" изображением, компетенции не имеется, чтоб разобраться в таких премудростях. Ну и буду следить за проектом, может что-то изменится со временем, проект отличный)

**Comment by xrip** [comment]
Для сеги возможно использовать денди джойстик как трехкнопочные джойстк Sega Megadrive 1, как сейчас сделано, где start = a, b = b, a = c

**Comment by DrNorfolk** [comment]
Да, я пробовал так, но червяку Джиму так неудобно прыгать на кнопку START )

**Comment by theavatar** [comment]
Hi xrip

I've been using compiling this and running on RP2040 zeros along with your other emulators, and have to say you've done a great job with them. I have modified some code to suit my needs but am facing the same issue above using a 2.8 / 2.6 TFT screen ( any ILI9342 at 240 * 320) and getting clipping on top and bottom of screen as well. I have modified the screen size variables which correctly resizes, however it is likely due to buffer size or sync rows issues that it is garbling the output image. If you have any ideas on this would be great. Otherwise, very happy with the speed and output.
